### ISSUE TITLE
error occurred when using <ranges>

### ISSUE THREAD

**Issue by stark-zhang** [clangd/clangd#873]
the error is following, you could reproduce it by following code in vscode. clangd seems cannot understand the `<ranges>` in C++ 20. 

![image](https://user-images.githubusercontent.com/17326718/134777866-2e3d723c-7957-4486-9f63-8d6d764967fe.png)
![image](https://user-images.githubusercontent.com/17326718/134778116-f398597b-397d-4df8-bb5a-d6be331a31a9.png)

the test code is following:
```c++
#include <ranges>
#include <iostream>
#include <utility>
#include <iomanip>
#include <type_traits>
#include <unordered_map>
#include <string>
#include <map>

using namespace std::ranges;

int main(int argc, char** argv)
{
    std::unordered_map<const char*, int> xmap = {
        {"James T. Kirk", 12},
        {"Spock", 20},
        {"Bones", -1},
    };

    for (const char* k : xmap | views::keys)
    {
        std::cout << std::setw(10) << k << std::endl;
    }
}

```
the project configuration:
```yaml
CompileFlags:
    Add: [-xc++, -std=c++20]
```

**System information**

Output of `clangd --version`:
`clangd 12.0.1`

Editor/LSP plugin:
vim/coc-nvim/coc-clangd
vscode/vscode-clangd
all of them is latest version.

Operating system:
OpenSUSE Tumbleweed 2021.08.27
Windows 10 1909 Enterprise

**Comment by HighCommander4** [comment]
I can confirm I see the same errors.

However, the errors are not specific to clangd: I see them when trying to compile this code with clang itself.

So, the issue seems to be clang not handling libstdc++'s <ranges> implementation. That could be a bug on the clang side, or a bug on the libstdc++ side -- hard to say without reducing the test case to understand the error.

**Comment by stark-zhang** [comment]
yes, I used `g++` to compile my code. this error will disappear if use the implementation of STL from LLVM, instead of GCC, won't it? It's hardly to me to find the bug in `libstdc++`., could you raise an issue to GCC?

**Comment by HighCommander4** [comment]
> Filed https://gcc.gnu.org/bugzilla/show_bug.cgi?id=102807

Based on the discussion in the libstdc++ issue, the bug is on the clang side, tracked in https://bugs.llvm.org/show_bug.cgi?id=44833.
